### PR TITLE
[18xx] Redridge Drudger

### PR DESCRIPTION
**Pull request by Phatcat**
Should know dual wield. (ref; https://github.com/cmangos/classic-db/issues/129)

Use addon system to set dual wield aura on spawn.

### CODE REVIEW

**Comment by killerwife** [comment]
I hate to rain on your parade, but sadly right now dual weild is only implemented for players. Does this in any way impair this PR? Or is it just visual and what-not nonsense?

**Comment by Phatcat** [comment]
What do you mean dual wield is only implemented for players?

Have you had a look in any of the ACID files..?

> '1104302','11043','1','0','100','0','1000','1000','0','0','11','674','0','0','0','0','0','0','0','0','0','0','Crimson Monk - Cast Dual Wield on Aggro'),

And that's just one of a good handful of AIs casting that very same dual wield aura on spawn.... and also...
billede

and what exactly do you even mean by 'your parade'?

**Comment by killerwife** [comment]
void Spell::EffectDualWield(SpellEffectIndex /*eff_idx*/)
{
    if (unitTarget && unitTarget->GetTypeId() == TYPEID_PLAYER)
        ((Player*)unitTarget)->SetCanDualWield(true);
}

**Comment by Phatcat** [comment]
But yes I think dual wield for mobs is a purely visual thing to get them swinging with both arms instead of just the one, I don't think they should actually have main damage and off damage no.

Well as you can quite clearly see, a lot of mobs in ACID already have dual wield set, and this mob should also evidently also know it (if you go visit him in his cave in-game, you can see that he is also holding 2 weapons, one in each hand, and currently only hitting with one of them) - I don't know if them knowing this aura will make him hit with both hands. It should... but if it doesn't, then all AIs who have dual wield set must be broken atm.... And why was it then never discovered by any of the other developers when implementing dual wield..? - either way if it works core wise or not, this mob should under all circumstances have it set, so irregardless, this PR is still valid.

**Comment by killerwife** [comment]
¯\_(ツ)_/¯

**Comment by Grz3s** [comment]
Explain this customer :)
has no dual implemented in acid and no related aura in addon.. and still works like this:
https://www.youtube.com/watch?v=ExbKTT8qK1c&feature=youtu.be

**Comment by Phatcat** [comment]
Well, something is up, because this is the redridge drudger, with or without this PR applied; https://youtu.be/enksPk9gCQc

I went to visit a syndicate assassin and indeed... he dual wields... without having dual wield (set as a spell/aura)... which he actually should (know dual wield as a spell that is, otherwise it wouldn't be listed here); 
billede


so... the situation is that dual wield is not extended to mobs in the core atm, making it broken using the correct method of giving them the dual wield aura, but at least one mob (probably others as well), this assassin, is using both his weapons, enabled by what I can then only guess is some flag or the like set in the creature_template... Anything out of the ordinary for the syndicate assassin @Grz3s ?

So you're right @killerwife , dual wield currently doesn't work for mobs in cmangos (unless they're hackfixed in by going around the dedicated dual wield aura..)

**Comment by Warlockbugs** [comment]
Creature spell books do not exist yet, so there is no capability of giving out passives for the time being, its done via database at the moment.

**Comment by Phatcat** [comment]
> Creature spell books do not exist yet, so there is no capability of giving out passives for the time being

Ehm... so the template addon and casting something on spawn/aggro in ACID can't/shouldn't be used for passives? Because... Yes, Dual wield is a passive... but so is something like Thrash... and we seem to be able to give mobs Thrash just fine using either ACID or template addon or even sd2. Why would this need a creature spell book system in order to work and not 'simply' for SetCanDualWield(true); and other such functionalities to be moved to Unit instead of Player, and then the check to see if it's a player, removed as well.. That would have to be done nomatter if the mob has a spellbook or not, because it'd still be the dual wield spell going into it, which would still be crippled by the core currently only accepting players as being able to dual wield.

**Comment by killerwife** [comment]
TL;DR
Add it to where you wanted to add it. Right now we simulate based on creature_equip_template, its sniffcandy only anyway.

**Comment by Phatcat** [comment]
> Right now we simulate based on creature_equip_template

What if there's a mob without any equipment which should know Dual Wield? like the Muckshell Pincer?

billede

billede

**Comment by killerwife** [comment]
Well you clearly know more than I do at this point.

**Comment by Warlockbugs** [comment]
Trying to add a passive effect (not an aura effect) into aura table will do exactly nothing, because such spells are working through spellbook only (for players), so commit itself is pointless.

Currently some abilities like that (parry, block) for creatures are governed via database flags:
CREATURE_EXTRA_FLAG_NO_PARRY
CREATURE_EXTRA_FLAG_NO_BLOCK
Which replace passive ability mechanism by imitating lack of "Parry" and "Block" abilities. Not having these flags imitates presence of these abilities on a mob.

The correct global workaround core-side would involve introducing a new flag:
CREATURE_EXTRA_FLAG_DUAL_WIELD

How are you going to do this in database - either abandoning equipment detection and migrating to a flag globally or use it as additional marker for creatures which have no equipment is up to database maintainers.

Removing all these hacks requires creature spellbooks, so we can account passives on mobs properly.

**Comment by Phatcat** [comment]
> Trying to add a passive effect (not an aura effect) into aura table will do exactly nothing, because such spells are working through spellbook only (for players), so commit itself is pointless.

That's a very convoluted way of saying 'you should rather set Dual Wield through ACID cause it's not an aura at all' - Sure, the commit itself is pointless, but the PR and this whole conversation far far from it.

> How are you going to do this in database - either abandoning equipment detection and migrating to a flag globally or use it as additional marker for creatures which have no equipment

What are you saying, that there should be 2 custom work-around systems? one for mobs with equipment and one for mobs without equipment?- currently, using the method of detecting it via equipment is bad, as it can't account for mobs without any equipment. The solution is not to continue as usual and just introduce a custom flag on top of that.... Sure, remove the equipment detection and only use a custom flag... but this is still custom. I'd prefer to find the method that blizzard used (which quite clearly involved the dual wield spell.....)

> How are you going to do this in database - ... is up to database maintainers.

So cmangos is not about doing emulation right?, it's about doing whatever hack the devs think is best..?

>Removing all these hacks requires creature spellbooks

Maybe. But it also requires for the spell effect Dual Wield to not check if the unit is a player, and then drop the unit if it isn't. In fact, if that is 'fixed' so that mobs can have dual wield properly, having them cast it through ACID, that combined with a unit bool 'is able to dual wield' or something may just as well be as optimal as any of those hacks you mentioned.

**Comment by killerwife** [comment]
Yeah, but despite your logic being sound, the problem is trying to solve all demons at once. I suggest you attempt what you wish for, but you might find its a pretty tall order to do everything at once "nicely".

**Comment by Phatcat** [comment]
Yeah, I have something concrete in mind I'm gonna try for this, I am gonna open a new PR later today when I get off work.

**Comment by Phatcat** [comment]
Still not currently supported by the core, though.
